### PR TITLE
Task done Withdrawal Address.

### PR DESCRIPTION
**Pull request by Momen123**
run this query before.

ALTER TABLE `users` ADD `withdrawalAddress` VARCHAR(35) NULL DEFAULT NULL AFTER `updatedTs`;

### CODE REVIEW

**Comment by thorie7912** [comment]
Looks good so far! Let's add a DB upgrade script for the SQL. Please read these instructions very carefully, and try to add the script: https://github.com/thorie7912/savings/blob/master/sql/README.md

Let me know if you have any issues.

File: app/Controller/IndexController.php [hunk 1]
```diff
@@ -13,29 +13,29 @@ class IndexController extends ViewController
     /** @var Session */
     private $session;
     private $loginForm;
-    
+
```
Comment (thorie7912) [inline app/Controller/IndexController.php]: Let's run the `cs-fix` alias so we don't have to worry about formatting changes.

File: app/Model/Menu.php [hunk 2]
```diff
@@ -54,7 +54,7 @@ public function getLinks()
 //            ];
             
             $links[] = [
-                'route' => 'user/settings',
+                'route' => 'settings/update-withdrawal-address',
```
Comment (thorie7912) [inline app/Model/Menu.php]: Let's use `settings/index` as the page with the form, and `settings/update-withdrawal-address` as the form post only. Let me know if this is unclear.

File: app/Controller/SettingsController.php [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+            $this->session->set('withdrawalAddressFormState', $this->updateWithdrawalAddressForm->getState());
+            $this->headers->redirect('settings/update-withdrawalAddress');
+            return;
+        }
+
+
+
+        $user = $this->createUserObject();
+        $user->id = $this->session->get('userId');
```
Comment (thorie7912) [inline app/Controller/SettingsController.php]: Let's use `$user->init(['id' => $this->session->get('userId')]);` instead of the above line. That way, later, when we call `saveUser()` it won't think that `id` has been modified. See my comment in the `Users` class.

File: app/Model/Users.php [hunk 4]
```diff
@@ -93,7 +94,7 @@ public function saveUser(User $user)
         
         $set = [];
         $params = [];
-        $typesMap = ['email' => 's', 'verifyCode' => 's', 'password' => 's', 'passwordResetCode' => 's'];
+        $typesMap = ['id' => 's', 'withdrawalAddress' => 's', 'email' => 's', 'verifyCode' => 's', 'password' => 's', 'passwordResetCode' => 's'];
```
Comment (thorie7912) [inline app/Model/Users.php]: We shouldn't need `id` here, if we make the proper change in the SettingsController as I commented.

File: app/Util/Validator.php [hunk 5]
```diff
@@ -43,4 +43,29 @@ public function isValidVerifyCodeString($verifyCode)
...
     }
+
+    //Momen
+    public function isValidBitcoinAddress($withdrawalAddress)
+    {
+        $available_chars = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';
+
+//        echo "here".$withdrawalAddress;
+
+        if(strlen($withdrawalAddress) >= 25 && strlen($withdrawalAddress) <= 34 && strpos ( $withdrawalAddress, '1' ) === 0 )
```
Comment (thorie7912) [inline app/Util/Validator.php]: Let's use class constants for these numbers 25 and 34, I suggest something like `self::MAX_BITCOIN_ADDRESS_LENGTH`.

File: app/Util/Validator.php [hunk 6]
```diff
@@ -43,4 +43,29 @@ public function isValidVerifyCodeString($verifyCode)
         return strlen($verifyCode) > 1 && strlen($verifyCode) < 100 &&
             strlen($verifyCode) == strspn($verifyCode, self::HEX);
     }
+
+    //Momen
+    public function isValidBitcoinAddress($withdrawalAddress)
+    {
+        $available_chars = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';
```
Comment (thorie7912) [inline app/Util/Validator.php]: Let's use a constant for this. And, since there are fewer disallowed characters, let's do something like `self::DISALLOWED_CHARS_REGEX = '/[<put bad chars here>]/'`. Then, we can check that the string has any of these bad characters anywhere in the string by searching with `$hasBadChars = preg_match('/' . self::DISALLOWED_CHARS . '/', $withdrawalAddress)`. It will return `true` if it finds a bad character.

Then, we don't need to create our own loop below to search.

File: app/View/user/settings.phtml [hunk 7]
```diff
@@ -1,6 +1,23 @@
...
+            <div style='color:red'><?=$vars['formErrors']['withdrawalAddress']?></div>
+        </div>
+    </div>
+    <div class="control-group">
+        <div class="controls">
+            <button type="submit" class="btn btn-primary">Update Bitcoin Address</button>
+        </div>
+    </div>
+</form>
+
```
Comment (thorie7912) [inline app/View/user/settings.phtml]: Let's move this whole file to `app\Views\settings\index.phtml` to support the view route that I commented about (since it's no longer part of the user controller).

File: app/Controller/SettingsController.php (hunk 3 above)
Comment (Momen123) [reply app/Controller/SettingsController.php]: it's not working like this.

Comment (Momen123) [reply app/Controller/SettingsController.php]: sorry, worked.

File: app/Model/Form/UpdateWithdrawalAddress.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+        $this->withdrawalAddress = $params['withdrawalAddress'];
+
+        if (!$valid_withdrawal_address) {
+
+            $this->setError('withdrawalAddress', 'Invalid withdrawal address.');
+        }
+//        else {
+//            $this->withdrawalAddress = $params['withdrawalAddress'];
+//        }
```
Comment (thorie7912) [inline app/Model/Form/UpdateWithdrawalAddress.php]: Do we need this comment?

File: app/Model/User.php [hunk 9]
```diff
@@ -27,6 +28,9 @@ public function __construct(Config $config, Email $mailer)
            'password',
            'verifyCode',
            'passwordResetCode',
+            //Momen
+            'withdrawalAddress'
+            //End Momen
```
Comment (thorie7912) [inline app/Model/User.php]: Do we need this comment?

File: app/Controller/SettingsController.php [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+        if(isset($user->withdrawalAddress))
+        {
+            $defaultState['CwithdrawalAddress'] = $user->withdrawalAddress;
+        }
+        else
+        {
+            $defaultState['CwithdrawalAddress'] = '';
+        }
+        $this->view->addVars($defaultState);
```
Comment (thorie7912) [inline app/Controller/SettingsController.php]: Instead of adding this as a custom value, we should use the form.
So, we can do `$this->updateWithdrawalAddressForm->withdrawalAddress = $user->withdrawalAddress` and it will get loaded into the form.

File: app/Model/Form/UpdateWithdrawalAddress.php [hunk 11]
```diff
@@ -0,0 +1,46 @@
+<?php
+
+//Momen
```
Comment (thorie7912) [inline app/Model/Form/UpdateWithdrawalAddress.php]: Extra comment

File: app/Model/Form/UpdateWithdrawalAddress.php [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+            $this->setError('withdrawalAddress', 'Invalid withdrawal address.');
+        }
+//        else {
+//            $this->withdrawalAddress = $params['withdrawalAddress'];
+//        }
+    }
+}
+
+//End Momen
```
Comment (thorie7912) [inline app/Model/Form/UpdateWithdrawalAddress.php]: Extra comment

File: app/Model/Users.php [hunk 13]
```diff
@@ -73,9 +73,10 @@ public function loadUser(User $user)
             $types = 'i';
             $params = [$user->id];
         }
-        
-        $query = "select id, email, verifyCode, passwordResetCode from users where $where limit 1";
-        
+
+        //Momen added the withdrawal_address to yhe query
+        $query = "select id, email, verifyCode, passwordResetCode, withdrawalAddress from users where $where limit 1";
+        //End Momen
```
Comment (thorie7912) [inline app/Model/Users.php]: Extra comment

File: app/Util/Validator.php [hunk 14]
```diff
@@ -43,4 +43,11 @@ public function isValidVerifyCodeString($verifyCode)
         return strlen($verifyCode) > 1 && strlen($verifyCode) < 100 &&
             strlen($verifyCode) == strspn($verifyCode, self::HEX);
     }
+
+    //Momen
+    public function isValidBitcoinAddress($withdrawalAddress)
+    {
+        return preg_match(self::BITCOIN_ADDRESS_PATTEERN_REGEX, $withdrawalAddress);
+    }
+    //End Momen
```
Comment (thorie7912) [inline app/Util/Validator.php]: Extra comment

**Comment by thorie7912** [comment]
Looks like good progress. Next, let's add some unit tests for all the new code, and any code that has been modified. Are you able to run the test suite locally?

File: app/EmailTemplate/verify-code.tpl [hunk 15]
```diff
@@ -1,4 +1,4 @@
-Thank you for signing up with Savings for Bitcoin!
+Thank you for signing up with Bitvest!
```
Comment (thorie7912) [inline app/EmailTemplate/verify-code.tpl]: @Momen123 I think your way of using git has a problem. You shouldn't be overwriting changes other people made already.

Can you be sure to use git push+pull instead of directly uploading files from the site?

File: sql/upgrade/fbddc9a64cb4e4cd1c78ef41f5e6cd72.sql [hunk 16]
```diff
@@ -0,0 +1,5 @@
+-- issue #6 Allow the user to add a payout Bitcoin address
```
Comment (thorie7912) [inline sql/upgrade/fbddc9a64cb4e4cd1c78ef41f5e6cd72.sql]: I think this filename (hash) is not correct.
I think it should be `263e3f361910448092ca733f96863f9f.sql`. Perhaps you have some other schema modifications that are not expected? Or maybe you took the hash AFTER adding the column, instead of before?

File: app/Model/Auth.php [hunk 17]
```diff
@@ -32,21 +32,4 @@ public function checkPassword($email, $password)
...
-            'id' => $id,
-        ];
-        $result = $this->mysql->query($query, 'i', $params);
-        
-        if (empty($result)) {
-            return false;
-        }
-        
-        return password_verify($password, $result[0]['password']);
-    }
```
Comment (thorie7912) [inline app/Model/Auth.php]: Why is this removed?

File: app/View/user/settings.phtml [hunk 18]
```diff
@@ -1,10 +1,24 @@
...
 
 <p>update payment method, change password, enable 2FA, delete account</p>
-<br>
-<h4 class="muted">&raquo; Change Password</h4>
-<?php include('change-password-form.phtml'); ?>
```
Comment (thorie7912) [inline app/View/user/settings.phtml]: Why is this removed?

File: app/Model/Users.php [hunk 19]
```diff
@@ -35,7 +35,9 @@ public function add(User $user)
                 . '(email, password, verifyCode, createdTs, updatedTs) '
                 . 'values (?, ?, ?, NOW(), NOW())';
         
-        $params = [$user->email, $user->password, $verifyCode];
+        $hash = password_hash($user->password, PASSWORD_BCRYPT);
+        
+        $params = [$user->email, $hash, $verifyCode];
```
Comment (thorie7912) [inline app/Model/Users.php]: This was not merged from `master` branch correctly. On `master` the code was changed to NOT perform `password_hash()` here. Your merged result is undo-ing that work.

File: app/Model/Users.php [hunk 20]
```diff
@@ -71,9 +73,8 @@ public function loadUser(User $user)
             $types = 'i';
             $params = [$user->id];
         }
-        
-        $query = "select id, email, password, verifyCode, passwordResetCode from users where $where limit 1";
-        
+
+        $query = "select id, email, verifyCode, passwordResetCode, withdrawalAddress from users where $where limit 1";
```
Comment (thorie7912) [inline app/Model/Users.php]: This was not merged correctly. The `master` branch was modified to select `password`. This merge is undo-ing that work.

File: app/Model/Users.php [hunk 21]
```diff
@@ -88,16 +89,21 @@ public function saveUser(User $user)
...
+        $typesMap = ['withdrawalAddress' => 's', 'email' => 's', 'verifyCode' => 's', 'password' => 's', 'passwordResetCode' => 's'];
         $types = [];
         
         foreach ($user->getModifiedProperties() as $property) {
             $set[] = "$property = ?";
             $types[] = $typesMap[$property];
-            $params[] = $user->$property;
+            
+            if ($property == 'password') {
+                $params[] = password_hash($user->password, PASSWORD_BCRYPT);
```
Comment (thorie7912) [inline app/Model/Users.php]: This merge is incorrect. Please see `master` branch. It does not use `password_hash()` anymore.

File: tests/app/Controller/SettingsControllerTest.php [hunk 22]
```diff
@@ -0,0 +1,139 @@
+<?php
+namespace App\Controller;
```
Comment (thorie7912) [inline tests/app/Controller/SettingsControllerTest.php]: This was not merged correctly. The `master` branch doesn't have a `tests/app` folder anymore.

File: tests/app/Controller/UserControllerTest.php [hunk 23]
```diff
@@ -1,8 +1,7 @@
 <?php
-namespace Test\App\Controller;
+namespace App\Controller;
```
Comment (thorie7912) [inline tests/app/Controller/UserControllerTest.php]: This was not merged correctly. The `master` branch doesn't have a `tests/app` folder anymore.

File: tests/app/Model/Form/UpdateWithdrawalAddressTest.php [hunk 24]
```diff
@@ -0,0 +1,69 @@
+<?php
+namespace Test\App\Model\Form;
```
Comment (thorie7912) [inline tests/app/Model/Form/UpdateWithdrawalAddressTest.php]: This test should be moved to a folder without `app`, e.g. `tests/Model/Form/UpdateWithdrawalAddressTest.php`.

File: tests/app/Model/Form/WithUpdateWithdrawalAddressExpectations.php [hunk 25]
```diff
@@ -0,0 +1,29 @@
+<?php
```
Comment (thorie7912) [inline tests/app/Model/Form/WithUpdateWithdrawalAddressExpectations.php]: Moved, same as others without `app/` in the folders.

File: tests/app/Util/ValidatorTest.php [hunk 26]
```diff
@@ -20,8 +20,11 @@ public function testValidation()
```
Comment (thorie7912) [inline tests/app/Util/ValidatorTest.php]: Moved, same as others, without `app/` in the path.
